sql: add a new supported event to the conn state machine

andreimatei opened the pull request:
This patch adds support for the event generated when running a `rollback
to savepoint cockroach_restart` when in state Open (so, when no error
has occured). We support running this with the current Executor code;
this adds feature parity to the conn_fsm.

Release note: None

andreimatei committed:
sql: add a new supported event to the conn state machine

This patch adds support for the event generated when running a `rollback
to savepoint cockroach_restart` when in state Open (so, when no error
has occured). We support running this with the current Executor code;
this adds feature parity to the conn_fsm.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22368)
<!-- Reviewable:end -->

andreimatei commented:
The first commit is #22367

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r1.
Review status: 3 of 8 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-:-L4c2Tq_5nsC-MzWsgaA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Wait no
I didn't review this!

knz commented:
I was looking at #22367 and lgtmed here by accident. Looking here now.

knz commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/conn_fsm.go, line 343 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3Br1llWsRYHL6HRQ:-L4c3Br23mmENSqqzSRz:blbgk6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/sql/conn_fsm.go#L343)):*
> ```Go
> 				// the client txn's proto to increment the epoch.
> 				if state.mu.txn.GetTxnCoordMeta().CommandCount > 0 {
> 					// NOTE: We don't bump the txn timestamp on this restart. If this
> ```

extraneous "this"

---

*[pkg/sql/conn_fsm.go, line 347 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3GkcIvcwSIcfLS3m:-L4c3Gkd_gta3vOvTbVJ:bysmvuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/sql/conn_fsm.go#L347)):*
> ```Go
> 					// a regular savepoint, clearly we couldn't bump the timestamp. But in
> 					// the special case of the cockroach_restart savepoint, it's not clear
> 					// to me what a user's expectation might be.
> ```

I don't understand this comment. What is this telling me?

I would like a comment that tells me why the restart is conditional on CommandCount (and also why it's not correct to restart even when the count is 0).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 2 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/sender.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3o3AnhuiwUrLjWRu:-L4c3o3BMui_chYT5_nY:b-b34x10) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/internal/client/sender.go#L105)):*
> ```Go
> 
> // TxnSenderFunc is an adapter to allow the use of ordinary functions
> // as TxnSenders with GetMeta or AugmentMeta panicing with unimplemented.
> ```

This is now incorrect.

---

*[pkg/sql/conn_fsm.go, line 341 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c4uFzbymQoR_x02IK:-L4c4uFzbymQoR_x02IL:b-lxe16s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/sql/conn_fsm.go#L341)):*
> ```Go
> 				state := args.Extended.(*txnState2)
> 				// If commands have already been sent through the transaction, restart
> 				// the client txn's proto to increment the epoch.
> ```

Why not do this unconditionally?

---

*[pkg/sql/conn_fsm.go, line 347 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3GkcIvcwSIcfLS3m:-L4c4GiGClYkt5L-WC6C:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/sql/conn_fsm.go#L347)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't understand this comment. What is this telling me?

I would like a comment that tells me why the restart is conditional on CommandCount (and also why it's not correct to restart even when the count is 0).
</blockquote></details>

+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/sender.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3o3AnhuiwUrLjWRu:-L4nR1thfnCpxiqZcFbe:bypydyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/internal/client/sender.go#L105)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is now incorrect.
</blockquote></details>

it was correct in spirit :)
Expanded.

---

*[pkg/sql/conn_fsm.go, line 341 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c4uFzbymQoR_x02IK:-L4nQ-IGilZv9GvrorXl:b-wpxdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/sql/conn_fsm.go#L341)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why not do this unconditionally?
</blockquote></details>

because it isn't necessary. But switched to doing it unconditionally.

---

*[pkg/sql/conn_fsm.go, line 343 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3Br1llWsRYHL6HRQ:-L4nPvgDwOX5xa7tmdup:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/sql/conn_fsm.go#L343)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

extraneous "this"
</blockquote></details>

Done.

---

*[pkg/sql/conn_fsm.go, line 347 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3GkcIvcwSIcfLS3m:-L4nQMYfrx1KyC14I33k:btnr7sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/sql/conn_fsm.go#L347)):*
> why it's not correct to restart even when the count is 0

It is correct, just not necessary. I've switched to always doing it now.

This comment highlights the fact that we don't bump the timestamp, and then discusses whether we should or not. Added more elements of stye to hopefully make it clear.
Rhetoric aside, I'd prolly bump it if I had a clock around, but alas I don't. It was the same way in the `Executor` code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/sender.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3o3AnhuiwUrLjWRu:-L4qPJWjOyn1IIOkJ-a5:b-fz735g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/internal/client/sender.go#L105)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it was correct in spirit :)
Expanded.
</blockquote></details>

Does anything in testing use this now? Can we make it a panic again? That was one of the reasons why I wanted to do the txn restart unconditionally.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

have nathan green light this too

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-:-L4vxEhNibgwkxtcXUhq:bbq52ri)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/internal/client/sender.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-L4c3o3AnhuiwUrLjWRu:-L59a0loNJPhF9fUlRPG:b-ky38rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1213ef535f6cbd082f12b0361a6a25b5a7fcbf01/pkg/internal/client/sender.go#L105)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does anything in testing use this now? Can we make it a panic again? That was one of the reasons why I wanted to do the txn restart unconditionally.
</blockquote></details>

good point. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22368#-:-L59b6DqcrrKSfgPng7Z:bnfp4nl)*
<!-- Sent from Reviewable.io -->
